File: sdks/typescript/src/decorators.ts
/**
 * Workflow and Step Decorators
 * TypeScript decorators for defining resumable workflows
 */

import { WorkflowConfig, StepConfig, RetryPolicy, WorkflowState } from './types';
import { ExecutionContext, ExecutionEngine, Lease } from './context';
import {
  WorkflowLocked,
  StepTimeout,
  TooManyAttempts,
  StepExecutionFailed,
} from './errors';
import { WorkflowRegistry } from './registry';

// ============================================================================
// Retry Policy Helpers
// ============================================================================

function shouldRetry(
  policy: RetryPolicy,
  attempt: number,
  error: Error
): boolean {
  if (attempt >= policy.maxAttempts) return false;
  if (policy.retryableErrors && policy.retryableErrors.length > 0) {
    return policy.retryableErrors.includes(error.name);
  }
  return true;
}

function calculateBackoff(policy: RetryPolicy, attempt: number): number {
  const delay = Math.min(
    Math.pow(policy.backoffBase, attempt),
    policy.backoffMax
  );
  const jitterRange = delay * policy.backoffJitter;
  return delay - jitterRange / 2 + Math.random() * jitterRange;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ============================================================================
// Workflow Decorator
// ============================================================================

export interface WorkflowOptions extends WorkflowConfig {
  engine?: ExecutionEngine;
}

/**
 * Mark a function as a resumable workflow
 * 
 * @example
 * ```typescript
 * @workflow({ tags: { team: 'platform' } })
 * async function processOrder(orderId: string) {
 *   await validateOrder(orderId);
 *   await chargePayment(orderId);
 *   await shipOrder(orderId);
 * }
 * ```
 */
export function workflow(options: WorkflowOptions = {}) {
  return function <T extends (...args: unknown[]) => Promise<unknown>>(
    target: T,
    context: ClassMethodDecoratorContext | undefined
  ): T {
    const workflowName =
      typeof context?.name === 'string' ? context.name : target.name;

    const wrapper = async function (
      this: unknown,
      ...args: Parameters<T>
    ): Promise<ReturnType<T>> {
      const startTime = Date.now();

      // Get or create context
      const ctx = ExecutionContext.getOrCreate(
        options.workflowId,
        workflowName,
        options.orgId,
        options.tags
      );

      const engine = options.engine;
      if (!engine) {
        throw new Error('ExecutionEngine not provided in workflow options');
      }

      ctx.setEngine(engine);

      // Acquire lease
      const leaseStart = Date.now();
      const lease = await engine.leaseManager.acquire(
        ctx.workflowId,
        ctx.executorId
      );

      if (!lease) {
        throw new WorkflowLocked(ctx.workflowId);
      }

      ctx.setLease(lease);

      try {
        // Start heartbeat
        ctx.startHeartbeat(lease, engine);

        // Check if resuming
        if (ctx.isResuming()) {
          const state = await engine.restore(ctx.workflowId);
          ctx.setState(state);
          console.log(
            `Resumed workflow ${ctx.workflowId} from step ${state.stepNumber}`
          );
        }

        // Execute workflow within context
        const result = await ctx.run(() => target.apply(this, args));

        // Mark complete
        await engine.completeWorkflow(ctx.workflowId);

        return result as ReturnType<T>;
      } finally {
        ctx.stopHeartbeat();
        await engine.leaseManager.release(lease);
      }
    } as T;

    // Attach metadata
    Object.defineProperty(wrapper, '__contd_workflow__', { value: true });
    Object.defineProperty(wrapper, '__contd_config__', { value: options });
    Object.defineProperty(wrapper, 'name', { value: workflowName });

    // Register workflow
    WorkflowRegistry.register(workflowName, wrapper);

    return wrapper;
  };
}

// ============================================================================
// Step Decorator
// ============================================================================

export interface StepOptions extends StepConfig {}

/**
 * Mark a function as a workflow step
 * 
 * @example
 * ```typescript
 * @step({
 *   retry: { maxAttempts: 3, backoffBase: 2 },
 *   timeout: 30000
 * })
 * async function chargePayment(orderId: string): Promise<{ paymentId: string }> {
 *   const result = await paymentGateway.charge(orderId);
 *   return { paymentId: result.id };
 * }
 * ```
 */
export function step(options: StepOptions = {}) {
  const config: Required<StepConfig> = {
    checkpoint: options.checkpoint ?? true,
    idempotencyKey: options.idempotencyKey ?? undefined,
    retry: options.retry ?? undefined,
    timeout: options.timeout ?? undefined,
    savepoint: options.savepoint ?? false,
  };

  return function <T extends (...args: unknown[]) => Promise<unknown>>(
    target: T,
    context: ClassMethodDecoratorContext | undefined
  ): T {
    const stepName =
      typeof context?.name === 'string' ? context.name : target.name;

    const wrapper = async function (
      this: unknown,
      ...args: Parameters<T>
    ): Promise<ReturnType<T>> {
      const ctx = ExecutionContext.current();
      const engine = ctx.getEngine();
      const lease = ctx.getLease();

      if (!engine) {
        throw new Error('No execution engine in context');
      }

      // Generate step ID
      const stepId = ctx.generateStepId(stepName);

      // Check idempotency
      const cachedResult = await engine.idempotency.checkCompleted(
        ctx.workflowId,
        stepId
      );

      if (cachedResult) {
        console.log(`Step ${stepId} already completed, returning cached result`);
        ctx.setState(cachedResult);
        return cachedResult as ReturnType<T>;
      }

      // Allocate attempt
      const attemptId = await engine.idempotency.allocateAttempt(
        ctx.workflowId,
        stepId,
        lease
      );

      // Write intention
      await engine.journal.append({
        eventId: crypto.randomUUID(),
        workflowId: ctx.workflowId,
        orgId: ctx.orgId,
        timestamp: new Date().toISOString(),
        eventType: 'step_intention',
        stepId,
        stepName,
        attemptId,
      });

      // Execute with timeout and retry
      const startTime = Date.now();
      let lastError: Error | null = null;
      let currentAttempt = attemptId;

      const executeStep = async (): Promise<unknown> => {
        if (config.timeout) {
          return executeWithTimeout(
            () => target.apply(this, args),
            config.timeout,
            ctx.workflowId,
            stepId,
            stepName
          );
        }
        return target.apply(this, args);
      };

      try {
        const result = await executeStep();
        const durationMs = Date.now() - startTime;

        // Extract new state
        const newState = ctx.extractState(result);
        const oldState = ctx.getState();

        // Compute delta (simplified)
        const delta = computeDelta(oldState, newState);

        // Write completion
        await engine.journal.append({
          eventId: crypto.randomUUID(),
          workflowId: ctx.workflowId,
          orgId: ctx.orgId,
          timestamp: new Date().toISOString(),
          eventType: 'step_completed',
          stepId,
          attemptId: currentAttempt,
          stateDelta: delta,
          durationMs,
        });

        // Mark completed
        await engine.idempotency.markCompleted(
          ctx.workflowId,
          stepId,
          currentAttempt,
          newState
        );

        // Update context
        ctx.setState(newState);
        ctx.incrementStep();

        // Checkpoint if configured
        if (config.checkpoint) {
          await engine.maybeSnapshot(newState);
        }

        // Savepoint if configured
        if (config.savepoint) {
          await ctx.createSavepoint();
        }

        return result as ReturnType<T>;
      } catch (error) {
        const durationMs = Date.now() - startTime;
        lastError = error as Error;

        // Log failure
        await engine.journal.append({
          eventId: crypto.randomUUID(),
          workflowId: ctx.workflowId,
          orgId: ctx.orgId,
          timestamp: new Date().toISOString(),
          eventType: 'step_failed',
          stepId,
          attemptId: currentAttempt,
          error: lastError.message,
        });

        // Check retry policy
        if (config.retry && shouldRetry(config.retry, currentAttempt, lastError)) {
          const backoff = calculateBackoff(config.retry, currentAttempt);
          console.log(`Retrying step ${stepId}, attempt ${currentAttempt + 1}`);
          await sleep(backoff * 1000);
          return wrapper.apply(this, args);
        }

        // Check max attempts
        if (config.retry && currentAttempt >= config.retry.maxAttempts) {
          throw new TooManyAttempts(
            ctx.workflowId,
            stepId,
            stepName,
            config.retry.maxAttempts,
            lastError.message
          );
        }

        // Wrap error
        throw new StepExecutionFailed(
          ctx.workflowId,
          stepId,
          stepName,
          currentAttempt,
          lastError
        );
      }
    } as T;

    // Attach metadata
    Object.defineProperty(wrapper, '__contd_step__', { value: true });
    Object.defineProperty(wrapper, '__contd_config__', { value: config });
    Object.defineProperty(wrapper, 'name', { value: stepName });

    return wrapper;
  };
}

// ============================================================================
// Helpers
// ============================================================================

async function executeWithTimeout<T>(
  fn: () => Promise<T>,
  timeoutMs: number,
  workflowId: string,
  stepId: string,
  stepName: string
): Promise<T> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      reject(
        new StepTimeout(
          workflowId,
          stepId,
          stepName,
          timeoutMs / 1000,
          timeoutMs / 1000
        )
      );
    }, timeoutMs);

    fn()
      .then((result) => {
        clearTimeout(timer);
        resolve(result);
      })
      .catch((error) => {
        clearTimeout(timer);
        reject(error);
      });
  });
}

function computeDelta(
  oldState: WorkflowState,
  newState: WorkflowState
): Record<string, unknown> {
  const delta: Record<string, unknown> = {};

  // Compare variables
  for (const [key, value] of Object.entries(newState.variables)) {
    if (JSON.stringify(oldState.variables[key]) !== JSON.stringify(value)) {
      delta[key] = value;
    }
  }

  // Check for removed keys
  for (const key of Object.keys(oldState.variables)) {
    if (!(key in newState.variables)) {
      delta[key] = null;
    }
  }

  return delta;
}

File: contd/sdk/decorators.py
from typing import TypeVar, Callable, ParamSpec, Optional
from dataclasses import dataclass
from functools import wraps
from datetime import timedelta
import time
import traceback
import logging

from contd.sdk.context import ExecutionContext, utcnow, generate_id
from contd.sdk.types import RetryPolicy
from contd.sdk.errors import WorkflowLocked
from contd.models.serialization import compute_delta
from contd.models.events import StepIntentionEvent, StepFailedEvent, StepCompletedEvent

logger = logging.getLogger(__name__)

P = ParamSpec('P')
R = TypeVar('R')

@dataclass
class WorkflowConfig:
    workflow_id: Optional[str] = None      # Auto-generate if None
    max_duration: Optional[timedelta] = None
    retry_policy: Optional[RetryPolicy] = None
    tags: Optional[dict[str, str]] = None

def workflow(config: WorkflowConfig | None = None):
    """
    Mark a function as a resumable workflow.
    """
    def decorator(fn: Callable[P, R]) -> Callable[P, R]:
        @wraps(fn)
        def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
            # Get or create workflow context
            ctx = ExecutionContext.get_or_create(
                workflow_id=config.workflow_id if config else None,
                workflow_name=fn.__name__,
                tags=config.tags if config else None
            )
            
            # Acquire lease
            lease = ctx.engine.lease_manager.acquire(
                ctx.workflow_id,
                owner_id=ctx.executor_id
            )
            
            if not lease:
                raise WorkflowLocked(f"Workflow {ctx.workflow_id} locked by another executor")
            
            try:
                # Start heartbeat background thread
                ctx.start_heartbeat(lease)
                
                # Check if resuming
                if ctx.is_resuming():
                    state = ctx.engine.restore(ctx.workflow_id)
                    ctx.set_state(state)
                    logger.info(f"Resumed workflow {ctx.workflow_id} from step {state.step_number}")
                
                # Execute workflow
                result = fn(*args, **kwargs)
                
                # Mark complete
                ctx.engine.complete_workflow(ctx.workflow_id)
                
                return result
                
            finally:
                ctx.stop_heartbeat()
                ctx.engine.lease_manager.release(lease)
        
        # Attach metadata for introspection
        wrapper.__contd_workflow__ = True
        wrapper.__contd_config__ = config
        
        return wrapper
    
    return decorator

@dataclass
class StepConfig:
    checkpoint: bool = True              # Create checkpoint after step?
    idempotency_key: Callable | None = None  # For external side effects
    retry: RetryPolicy | None = None
    timeout: timedelta | None = None
    savepoint: bool = False              # Rich savepoint with metadata?

def execute_with_timeout(fn, timeout: timedelta, *args, **kwargs):
    # Simplistic timeout implementation (real one might use signals or threads)
    # This is placeholder
    return fn(*args, **kwargs)

def step(config: StepConfig | None = None):
    """
    Mark a function as a workflow step.
    """
    cfg = config or StepConfig()
    
    def decorator(fn: Callable) -> Callable:
        @wraps(fn)
        def wrapper(*args, **kwargs):
            ctx = ExecutionContext.current()
            
            # Generate step ID (deterministic from function name + position)
            step_id = ctx.generate_step_id(fn.__name__)
            
            # Check idempotency: already completed?
            if cached_result := ctx.engine.idempotency.check_completed(
                ctx.workflow_id, 
                step_id
            ):
                logger.info(f"Step {step_id} already completed, returning cached result")
                # Need to update context state with cached result!
                # Cached result is WorkflowState.
                # If cached_result returns state, we set it.
                ctx.set_state(cached_result)
                return cached_result.variables # Return variables to user function usually?
                # Wait, the workflow code expects `fn` result.
                # If `fn` returns state/dict, we should return that.
                # `check_completed` returns `WorkflowState`.
                # We should probably return `state.variables` if user function expects dict, 
                # OR return `state` if users pass state around.
                # In examples: `state = search(state, query)`.
                # So we should return the appropriate object.
                # The prompt implies `state` (dict) is passed around. 
                # But `check_completed` returns `WorkflowState`.
                # For compatibility, this might need more robust typing or convention.
                # We'll assume the user code works with `variable` dicts if that's what `fn` does.
                # But the decorator expects `fn` to return something.
                # We'll return `cached_result` (WorkflowState)?
                # In example `state = fetch_order(state)`.
                # `fetch_order` takes state (dict or obj) and returns state.
                # If we return `WorkflowState`, user code might fail if it expects dict.
                # But `ExecutionContext.extract_state` handles both.
                # Let's return `cached_result` for now.
                return cached_result 

            # Allocate attempt ID
            attempt_id = ctx.engine.idempotency.allocate_attempt(
                ctx.workflow_id,
                step_id,
                ctx.lease
            )
            
            # Write intention
            ctx.engine.journal.append(StepIntentionEvent(
                event_id=generate_id(),
                workflow_id=ctx.workflow_id,
                timestamp=utcnow(),
                step_id=step_id,
                step_name=fn.__name__,
                attempt_id=attempt_id
            ))
            
            # Execute with timeout
            start_time = time.monotonic()
            try:
                if cfg.timeout:
                    result = execute_with_timeout(fn, cfg.timeout, *args, **kwargs)
                else:
                    result = fn(*args, **kwargs)
                    
            except Exception as e:
                # Log failure
                ctx.engine.journal.append(StepFailedEvent(
                    event_id=generate_id(),
                    workflow_id=ctx.workflow_id,
                    timestamp=utcnow(),
                    step_id=step_id,
                    attempt_id=attempt_id,
                    error=str(e), # spec used error_type/msg, but StepFailedEvent def in models uses `error: str`
                ))
                
                # Apply retry policy
                if cfg.retry and cfg.retry.should_retry(attempt_id, e):
                    logger.info(f"Retrying step {step_id}, attempt {attempt_id + 1}")
                    time.sleep(cfg.retry.backoff(attempt_id))
                    return wrapper(*args, **kwargs)  # Recursive retry
                
                raise
            
            duration_ms = int((time.monotonic() - start_time) * 1000)
            
            # Extract new state (assumes result is state or contains state)
            new_state = ctx.extract_state(result)
            old_state = ctx.get_state()
            
            # Compute delta
            # compute_delta expects dicts now based on my update to serialization.py
            delta = compute_delta(old_state.to_dict(), new_state.to_dict())
            
            # Write completion
            ctx.engine.journal.append(StepCompletedEvent(
                event_id=generate_id(),
                workflow_id=ctx.workflow_id,
                timestamp=utcnow(),
                step_id=step_id,
                attempt_id=attempt_id,
                state_delta=delta,
                duration_ms=duration_ms
            ))
            
            # Mark completed (idempotent)
            # Added last_event_seq=0 placeholder as I discussed in idempotency.py step
            ctx.engine.idempotency.mark_completed(
                ctx.workflow_id,
                step_id,
                attempt_id,
                new_state
            )
            
            # Update context state
            ctx.set_state(new_state)
            ctx.increment_step()
            
            # Checkpoint if configured
            if cfg.checkpoint:
                ctx.engine.maybe_snapshot(new_state)
            
            # Savepoint if configured (rich metadata)
            if cfg.savepoint:
                ctx.create_savepoint()
            
            return result
        
        wrapper.__contd_step__ = True
        wrapper.__contd_config__ = cfg
        
        return wrapper
    
    return decorator
